### PR TITLE
Strategy-specific E2E tests and edge cases for ExtractionPipeline. Closes #636

### LINKED ISSUE #636: Create end-to-end extraction pipeline tests
Although there are a some unit tests for the classes involved in the extraction process, the extraction pipeline as a whole is currently not tested end to end. So a comprehensive set of tests to cover the `ExtractionPipeline` class is needed. As these tests should also tackle the CowrieExtractionStrategy, it might be sensible to solve #634 first.

### PR DESCRIPTION
**Pull request by opbot-xd**
# Description

This PR adds comprehensive strategy-specific E2E tests and edge cases for the `ExtractionPipeline` class, completing the test coverage requested in issue #636.

Building on PR #735 (pipeline infrastructure + core flow tests), this PR adds:

### New Test Classes (24 tests total, ~795 lines):

1. **`TestCowrieStrategyE2E`** (3 tests)
   - Scanner extraction flow with session hits
   - Payload extraction from login messages with embedded URLs
   - File download URL extraction

2. **`TestLog4potStrategyE2E`** (3 tests)
   - JNDI/LDAP exploit extraction flow
   - Base64-encoded payload extraction with hidden URLs
   - Non-exploit hits filtering verification

3. **`TestGenericStrategyE2E`** (3 tests)
   - Fallback behavior for unknown honeypots
   - Heralding honeypot processing
   - Dionaea honeypot processing

4. **`TestMixedHoneypotE2E`** (1 test)
   - Mixed honeypot hits with correct strategy selection

5. **`TestEdgeCasesE2E`** (10 tests)
   - Malformed hits with missing required fields
   - Sensor extraction validation
   - Honeypot skipping when not ready (`is_ready_for_extraction` returning False)
   - Empty IOC records handling
   - Multiple strategy exceptions handling
   - Partial strategy success scenarios
   - Hit grouping verification
   - Special characters in fields
   - Large batch processing (1000 hits)

6. **`TestFactoryIntegration`** (4 tests)
   - Factory creates correct strategy for Cowrie
   - Factory creates correct strategy for Log4pot
   - Factory creates generic strategy for unknown honeypots
   - Case-sensitive honeypot name matching

## Related issues
Closes #636

## Type of change

- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowlproject.github.io/docs/GreedyBear/Contribute/) to this project.
- [x] The pull request is for the branch `develop`.
- [ ] I have added documentation of the new features.
- [x] Linter (`Ruff`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/GreedyBear/Contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved. All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/docs/blob/main/docs/GreedyBear/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### CODE REVIEW

**Comment by opbot-xd** [comment]
### Coverage

| Module | Coverage |
|--------|----------|
| `pipeline.py` | **100%** |
| `strategies/factory.py` | **100%** |
| `strategies/base.py` | **100%** |
| `strategies/generic.py` | **100%** |
| `strategies/__init__.py` | **100%** |

All 446 tests passing.

---

Hello @regulartim the PR is ready for review.
I believe the `test_extraction_pipeline.py` is getting huge (1183 lines). Any suggestions for that?

**Comment by opbot-xd** [comment]
Hi @regulartim 
My initial approach was unit testing the pipeline orchestration verifying that it correctly coordinates between factory, strategies, and scoring. I mocked the factory/strategies to isolate the pipeline logic from strategy implementation details.

However, you're right that for true E2E testing, we should let the real components run and only mock at the repository boundary. This gives us confidence that the actual integration works. I'll refactor to use the real factory and strategies.

File: tests/greedybear/cronjobs/test_extraction_pipeline_e2e.py [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+        """
+        with (
+            patch("greedybear.cronjobs.extraction.pipeline.SensorRepository"),
+            patch("greedybear.cronjobs.extraction.pipeline.IocRepository"),
+            patch("greedybear.cronjobs.extraction.pipeline.ElasticRepository"),
+        ):
+            from greedybear.cronjobs.extraction.pipeline import ExtractionPipeline
+
+            pipeline = ExtractionPipeline()
+            return pipeline
```
Comment (regulartim) [inline tests/greedybear/cronjobs/test_extraction_pipeline_e2e.py]: Does it make sense to move this to `tests/__init__.py`?

Comment (regulartim) [reply tests/greedybear/cronjobs/test_extraction_pipeline_e2e.py]: Or even the whole class?

Comment (opbot-xd) [reply tests/greedybear/cronjobs/test_extraction_pipeline_e2e.py]: Yes! this would make sense. Moving the whole class to `tests/__init__.py`.

**Comment by opbot-xd** [comment]
Hi @regulartim

Regarding the `TestIocContentVerification` class in `tests/greedybear/cronjobs/test_extraction_pipeline_e2e.py`:

We encountered a trade-off between using **real strategies** (as requested) and having **unconditional deterministic assertions**.

1.  **Copilot issue**: The tests use conditional assertions (`if mock_scores.return_value.score_only.called:`). This is technically non-deterministic because if extraction fails silently, the test still passes without verifying content.
2.  **The Conflict**: Guaranteeing IOC extraction with real strategies requires complex setup (specific hit fields, database state via patched repositories like `CowrieSessionRepository`). If we don't mock the factory to force an IOC return, we rely on the real strategy's logic.

**Decision**:
I have prioritized **using real production code**.
- The tests use real strategies and only mock the repositories (`CowrieSessionRepository`, `IocRepository`).
- The assertions remain conditional: *IF* the real strategy extracts IOCs (which it should, given the mock hits), *THEN* we verify the content strictly.

I believe this adheres best to the goal of "testing the actual integration path" while still adding the requested content verification.

**Comment by regulartim** [comment]
Hey @opbot-xd ! Thank you for your work. Are you done with working on this? To me it looks good now. Would like to approve and merge.

**Comment by opbot-xd** [comment]
Yes I have finished my work over this PR. Should I update the PR desc?

**Comment by regulartim** [comment]
No, all good! :)
